### PR TITLE
Shapes of nested lists [WIP -to be splitted in 3 separate PRs]

### PR DESCRIPTION
**Pull request by randomboolean**
- SHAPE
- HULLSHAPE
- CHECKSHAPE
- PERMUTE
- Update GET and PUT for nested lists.
- update FormattedWarpscriptFunction to check subtypes of Lists and Maps

### CODE REVIEW

File: warp10/src/main/java/io/warp10/script/formatted/FormattedWarpScriptFunction.java [hunk 1]
```diff
@@ -71,11 +71,31 @@ public ArgumentsBuilder addArgument(Class<?> clazz, String name, String doc) {
       return this;
     }
 
+    public ArgumentsBuilder addListArgument(Class<?> subClazz, String name, String doc) {
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/formatted/FormattedWarpScriptFunction.java]: This change should be in a separate PR as it is not solely linked to the SHAPE/tensor functions

File: warp10/src/main/java/io/warp10/script/formatted/ListSpecification.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
+//
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/formatted/ListSpecification.java]: Ditto, move to a separate PR

File: warp10/src/main/java/io/warp10/script/formatted/MapSpecification.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
+//
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/formatted/MapSpecification.java]: Ditto, move to a separate PR

File: warp10/src/main/java/io/warp10/script/functions/CHECKSHAPE.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+      if (el instanceof List) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  static Boolean recValidateShape(List list, List<Long> candidateShape) {
+    List<Long> copyShape =  new ArrayList<>(candidateShape);
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/CHECKSHAPE.java]: Use <Long> for clarity

File: warp10/src/main/java/io/warp10/script/functions/GET.java [hunk 5]
```diff
@@ -21,11 +21,12 @@
 import io.warp10.script.WarpScriptException;
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/GET.java]: Move the change to GET in a separate PR

File: warp10/src/main/java/io/warp10/script/functions/HULLSHAPE.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      Long y = taller.remove(0);
+      res.add(x > y ? x : y);
+    }
+
+    res.addAll(taller);
+    return res;
+  }
+
+  private List<Long> recHullShape(List list) {
+    List<Long> res = new ArrayList<>();
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/HULLSHAPE.java]: Use <Long> for clarity

File: warp10/src/main/java/io/warp10/script/functions/HULLSHAPE.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    res.add(new Long(list.size()));
+
+    List<Long> hull = new ArrayList<>();
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/HULLSHAPE.java]: Use <Long> for clarity

File: warp10/src/main/java/io/warp10/script/functions/PERMUTE.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
+package io.warp10.script.functions;
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/PERMUTE.java]: Missing copyright notice

File: warp10/src/main/java/io/warp10/script/functions/PERMUTE.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    pattern = (List) formattedArgs.get(PATTERN);
+    boolean fast = ((Boolean) formattedArgs.get(FAST)).booleanValue();
+
+    List<Long> shape = SHAPE.candidate_shape(tensor);
+
+    if (!(fast || CHECKSHAPE.recValidateShape(tensor, shape))) {
+      throw new WarpScriptException(getName() + " expects that the sizes of the nested lists are coherent together to form a tensor (or multidimensional array).");
+    }
+
+    newShape = new ArrayList<>();
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/PERMUTE.java]: Specify generic type for clarity

File: warp10/src/main/java/io/warp10/script/functions/PERMUTE.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    for (int r = 0; r < pattern.size(); r++) {
+      newShape.add(shape.get(pattern.get(r).intValue()));
+    }
+
+    List<Object> result = new ArrayList<>();
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/PERMUTE.java]: Add type for clarity

File: warp10/src/main/java/io/warp10/script/functions/PUT.java [hunk 11]
```diff
@@ -21,11 +21,12 @@
 import io.warp10.script.WarpScriptException;
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/PUT.java]: Move to a separate PR, maybe with the one containing the changes to GET

File: warp10/src/main/java/io/warp10/script/functions/GET.java [hunk 12]
```diff
@@ -85,4 +100,22 @@ public static int computeAndCheckIndex(NamedWarpScriptFunction func, int index,
 
     return index;
   }
+
+  static Object recNestedGet(NamedWarpScriptFunction func, List nestedList, List<Number> indexList) throws WarpScriptException {
+    List<Number> copyIndices = new ArrayList<>(indexList);
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/GET.java]: Add type for clarity

File: warp10/src/main/java/io/warp10/script/functions/GET.java [hunk 13]
```diff
@@ -85,4 +100,22 @@ public static int computeAndCheckIndex(NamedWarpScriptFunction func, int index,
 
     return index;
   }
+
+  static Object recNestedGet(NamedWarpScriptFunction func, List nestedList, List<Number> indexList) throws WarpScriptException {
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/GET.java]: Missing comment on what this method does

Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/GET.java]: This method should not need the func attribute, the WarpScriptException should be caught by the caller which will wrap it and add its name

File: warp10/src/main/java/io/warp10/script/functions/HULLSHAPE.java [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+  private List<Long> maximizeHull(List<Long> a, List<Long> b) {
+    List<Long> smaller = a.size() < b.size() ? a : b;
+    List<Long> taller = a.size() < b.size() ? b : a;
+
+    if (0 == smaller.size()) {
+      return taller;
+    }
+
+    List<Long> res = new ArrayList<>(taller.size());
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/HULLSHAPE.java]: Add type for clarity

File: warp10/src/main/java/io/warp10/script/functions/HULLSHAPE.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+      res.add(x > y ? x : y);
+    }
+
+    res.addAll(taller);
+    return res;
+  }
+
+  private List<Long> recHullShape(List list) {
+    List<Long> res = new ArrayList<>();
+    res.add(new Long(list.size()));
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/HULLSHAPE.java]: You can simply use res.add(list.size()), autoboxing will do the magic

File: warp10/src/main/java/io/warp10/script/functions/PERMUTE.java [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+      .build();
+  }
+
+  private List<Long> pattern;
+  private List<Long> newShape;
+
+  protected WarpScriptStack apply(Map<String, Object> formattedArgs, WarpScriptStack stack) throws WarpScriptException {
+    List<Object> tensor = (List) formattedArgs.get(TENSOR);
+    pattern = (List) formattedArgs.get(PATTERN);
+    boolean fast = ((Boolean) formattedArgs.get(FAST)).booleanValue();
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/PERMUTE.java]: Use Boolean.TRUE.equals(...)

File: warp10/src/main/java/io/warp10/script/functions/PERMUTE.java [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+      .addListArgument(Long.class, PATTERN, "The permutation pattern (a LIST of LONG).")
+      .addOptionalArgument(Boolean.class, FAST, "If True, it does not check if the sizes of the nested lists are coherent before operating. Default to False.", false)
+      .build();
+
+    output = new ArgumentsBuilder()
+      .addArgument(List.class, TENSOR, "The resulting nested LIST.")
+      .build();
+  }
+
+  private List<Long> pattern;
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/PERMUTE.java]: This should not be a field, as the apply method might be called from multiple threads

File: warp10/src/main/java/io/warp10/script/functions/PERMUTE.java [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  private void recPermute(List<Object> tensor, List<Object> result, List<Long> indices, int dimension) throws WarpScriptException {
+
+    for (int i = 0; i < newShape.get(dimension); i++) {
+      indices = new ArrayList(indices);
+      indices.add(new Long(i));
+
+      if (newShape.size() - 1 == dimension) {
+        List<Number> permutedIndices = new ArrayList<>();
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/PERMUTE.java]: Create List<Long> as we only use LONG in WarpScript for integer values

File: warp10/src/main/java/io/warp10/script/functions/SHAPE.java [hunk 19]
```diff
@@ -18,36 +18,62 @@
...
 
   @Override
-  public Object apply(WarpScriptStack stack) throws WarpScriptException {
-    Object top = stack.pop();
-    
-    if (!(top instanceof List)) {
-      throw new WarpScriptException(getName() + " operates on a list.");
+  protected WarpScriptStack apply(Map<String, Object> formattedArgs, WarpScriptStack stack) throws WarpScriptException {
+    List list = (List) formattedArgs.get(LIST);
+    boolean fast = ((Boolean) formattedArgs.get(FAST)).booleanValue();
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/SHAPE.java]: Use Boolean.TRUE.equals(...)

File: warp10/src/main/java/io/warp10/script/functions/SHAPE.java [hunk 20]
```diff
@@ -18,36 +18,62 @@
...
-    
-    List<Object> shape = new ArrayList<Object>();
-    
-    Object l = top;
-    
+    return stack;
+  }
+
+  static List<Long> candidate_shape(List list) {
+    List<Long> shape = new ArrayList<>();
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/SHAPE.java]: Add type for clarity

File: warp10/src/main/java/io/warp10/script/functions/SHAPE.java [hunk 21]
```diff
@@ -18,36 +18,62 @@
...
+    Object l = list;
+
     while(l instanceof List) {
-      shape.add(((List) l).size());
+      shape.add(new Long(((List) l).size()));
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/functions/SHAPE.java]: Let autoboxing do its job

File: warp10/src/test/java/io/warp10/script/NestedListsShapeTest.java [hunk 22]
```diff
@@ -0,0 +1,128 @@
+package io.warp10.script;
```
Comment (hbs) [inline warp10/src/test/java/io/warp10/script/NestedListsShapeTest.java]: Missing copytight notice

File: warp10/src/test/java/io/warp10/script/NestedListsShapeTest.java [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+import java.io.StringReader;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+
+public class NestedListsShapeTest {
+
+  private static final Path FILE_1 = Paths.get("src", "test", "java", "io", "warp10", "script", "nestedList_0.mc2");
```
Comment (hbs) [inline warp10/src/test/java/io/warp10/script/NestedListsShapeTest.java]: WarpScript files must be under src/test/warpscript

File: warp10/src/test/java/io/warp10/script/formatted/FormattedWarpScriptFunctionTest.java [hunk 24]
```diff
@@ -24,6 +24,7 @@
```
Comment (hbs) [inline warp10/src/test/java/io/warp10/script/formatted/FormattedWarpScriptFunctionTest.java]: Should be in a separate PR

File: warp10/src/test/java/io/warp10/script/nestedList_0.mc2 [hunk 25]
```diff
@@ -0,0 +1,103 @@
+       [  [  [  [   0   1   2 ]
```
Comment (hbs) [inline warp10/src/test/java/io/warp10/script/nestedList_0.mc2]: Move to src/test/warpscript and add copyright notice

File: warp10/src/test/java/io/warp10/script/nestedList_1.mc2 [hunk 26]
```diff
@@ -0,0 +1,82 @@
+      [ [ [ [   0   3   6   9 ]
```
Comment (hbs) [inline warp10/src/test/java/io/warp10/script/nestedList_1.mc2]: Move to src/test/warpscript and add copyright notice

File: warp10/src/main/java/io/warp10/script/functions/GET.java (hunk 13 above)
Comment (randomboolean) [reply warp10/src/main/java/io/warp10/script/functions/GET.java]: I think it needs the func attribute in case it is used by another function to have a better error message.

File: warp10/src/main/java/io/warp10/script/functions/PERMUTE.java (hunk 18 above)
Comment (randomboolean) [reply warp10/src/main/java/io/warp10/script/functions/PERMUTE.java]: permutedIndices is  used by GET.recNestedGet which can be applied to a List<Number>
(similar to GET that can be applied to Number, e.g. "[ 0 ] 0.3 GET" gives 0)

**Comment by randomboolean** [comment]
Was split between PR#468, PR#470, and PR#471
